src/object_store/cache.rs

```diff
@@ -217,88 +215,145 @@ impl CachingObjectStore {
-                            // Write task completed successfully, replace the in-memory cache entry
-                            // with the file-pointer one.
-                            debug!("Upserting file pointer for {key:?} into the cache");
-                            let value = CacheValue::File(path, size);
-                            cache.insert(key, value).await;
+        location: &object_store::path::Path,
+        start_chunk: usize,
+        end_chunk: usize,
+    ) -> object_store::Result<Bytes> {
+        let mut result = BytesMut::with_capacity(
+            (end_chunk.saturating_sub(start_chunk) + 1) * self.min_fetch_size as usize,
+        );
+
+        let mut chunk_batch = vec![];
+        for chunk in start_chunk..=end_chunk {
+            let chunk_range = (chunk * self.min_fetch_size as usize)
+                ..((chunk + 1) * self.min_fetch_size as usize);
+
+            let key = CacheKey {
+                path: location.to_owned(),
+                range: chunk_range.clone(),
+            };
+
+            let chunk_data = match self.cache.get(&key).await {
+                // If the value is missing extend the chunk range to fetch and continue
+                None => {
+                    chunk_batch.push(key);
+                    None
+                }
+                Some(value) => {
+                    // Now get the cache value for the current chunk
+                    match value {
+                        CacheValue::Memory(data) => {
+                            debug!("Cache value for {key:?} fetched from memory");
+                            Some(data.clone())
                         }
-                        Err(err) => {
-                            // Write task failed, remove the cache entry; we could also defer that to
-                            // TTL/LRU eviction, but then we risk ballooning the memory usage.
-                            warn!("Failed writing value for {key:?} to a file: {err}");
-                            warn!("Removing cache entry");
-                            cache.invalidate(&key).await;
+                        CacheValue::File(path, _) => {
+                            debug!("Cache value for {key:?} fetching from the file");
+                            match self.file_manager.read_file(path).await {
+                                Ok(data) => Some(data),
+                                Err(err) => {
+                                    warn!(
+                                        "Re-downloading cache value for {key:?}: {err}"
+                                    );
+                                    let data = self
+                                        .inner
+                                        .get_range(location, chunk_range.clone())
+                                        .await?;
+                                    self.cache_chunk_data(key, data.clone()).await;
+                                    Some(data)
```

Comment:
Worth adding a comment here to clarify the race condition this mitigates against:

thread 1 gets a value from cache, it's File(some-path) (self.cache.get on line 254)
thread 2 runs eviction, deletes the value and the file
thread 1 tries to read some-path, it's gone now

Reply:
It's a shame there isn't some get-with Moka functionality that takes in a closure that the value from getting evicted while it's running (dual to the comment above)